datafusion/physical-plan/src/union.rs

```diff
@@ -99,7 +99,12 @@ impl UnionExec {
     /// Create a new UnionExec
     pub fn new(inputs: Vec<Arc<dyn ExecutionPlan>>) -> Self {
         let schema = union_schema(&inputs);
-        let cache = Self::compute_properties(&inputs, schema);
+        // The schema of the inputs and the union schema is consistent when:
+        // - They have the same number of fields, and
+        // - Their fields have same types at the same indices.
+        // Here, we know that schemas are consistent and the call below can
+        // not return an error.
+        let cache = Self::compute_properties(&inputs, schema).unwrap();
```

Comment:
A slightly more defensive API might be to make this `try_new()` and return a `Result<Self>` rather than panic. However I think this is fine too

Reply:
I think it may be reasonable to return an `internal_err` as that would imply something that is not supposed to happen has happened. Let's think a little about doing that vs. panicking and do it in a follow-on PR if we decide to do so